### ISSUE TITLE
Planner produces wrong results if the generated plan contains same subplan in merge condition qual and sort key (derived by planner)

### ISSUE THREAD

**Issue by bhuvnesh2703** [greenplum-db/gpdb#2454]
This build is based on the PR: https://github.com/greenplum-db/gpdb/pull/2443 which eliminates subplan duplication in other cases.
Planner crashes (in debug build) if sort key and merge condition quals refers to the same subplan node, and produces wrong results in retail mode. 
```sql
CREATE TABLE foo (a integer, b integer)  PARTITION BY RANGE(b)
    (PARTITION sub_one START (1) END (10),
     PARTITION sub_two START (11) END (22));
CREATE TABLE bar (c integer, d character varying(10));
insert into foo values (9,9);
insert into bar values (9,9);
create table baz (i int4);
insert into baz values (9);

--Plan and Results (Retail Mode)

explain select bar.c from bar, foo where foo.b = (select max(i) from baz where bar.c = 9) and foo.b = bar.d::int4;
                                                             QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice3; segments: 3)  (cost=1000016899.07..1886953186.16 rows=18 width=4)
   ->  Merge Join  (cost=1000016899.07..1886953186.16 rows=6 width=4)
         Merge Cond: ((subplan)) = public.foo.b AND bar.d::integer = public.foo.b
         ->  Sort  (cost=1.07..1.08 rows=1 width=6)
               Sort Key: ((subplan))
               ->  Result  (cost=0.00..1.05 rows=1 width=6)
                     ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.05 rows=1 width=6)
                           ->  Seq Scan on bar  (cost=0.00..1.01 rows=1 width=6)
                     SubPlan 1
                       ->  Aggregate  (cost=5150.51..5150.52 rows=1 width=4)
                             ->  Result  (cost=0.00..1062.00 rows=32067 width=4)
                                   One-Time Filter: $0 = 9
                                   ->  Result  (cost=1158.20..2120.20 rows=32067 width=4)
                                         ->  Materialize  (cost=1158.20..2120.20 rows=32067 width=4)
                                               ->  Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..1062.00 rows=96200 width=4)
                                                     ->  Seq Scan on baz  (cost=0.00..1062.00 rows=32067 width=4)
         ->  Sort  (cost=16898.00..17328.50 rows=57400 width=4)
               Sort Key: public.foo.b
               ->  Append  (cost=0.00..1922.00 rows=57400 width=4)
                     ->  Seq Scan on foo_1_prt_sub_one foo  (cost=0.00..961.00 rows=28700 width=4)
                     ->  Seq Scan on foo_1_prt_sub_two foo  (cost=0.00..961.00 rows=28700 width=4)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
pivotal=# select bar.c from bar, foo  where foo.b = (select max(i) from baz where bar.c = 9) and foo.b = bar.d::int4;
 c
---
(0 rows)
```
Fails with debug build:
```sql
explain select bar.c from bar, foo where foo.b = (select max(i) from baz where bar.c = 9) and foo.b = bar.d::int4;
FATAL:  Unexpected internal error (nodeMotion.c:934)
DETAIL:  FailedAssertion("!(recvSlice->gangSize == 1)", File: "nodeMotion.c", Line: 934)
```

**Comment by bhuvnesh2703** [comment]
Results with ORCA
```sql
pivotal=# set optimizer=on;
SET
pivotal=# select bar.c from bar, foo  where foo.b = (select max(i) from baz where bar.c = 9) and foo.b = bar.d::int4;
 c
---
 9
(1 row)
```

**Comment by bhuvnesh2703** [comment]
Details of the patch for this fix available on https://github.com/greenplum-db/gpdb/pull/2443. Will update it once 2443 goes in.

**Comment by bhuvnesh2703** [comment]
Included the fix for it in #2443.

**Comment by hsyuan** [comment]
close this issue since it was fixed
